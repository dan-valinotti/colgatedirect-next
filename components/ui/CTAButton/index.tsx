import React, { FunctionComponent } from 'react';
import { Styled } from './_styles';

type Props = {
  /**
   * Color of button text and border.
   */
  color: string;
  /**
   * Text contained in button.
   */
  text: React.ReactNode;
  /**
   * Function to be executed on button click.
   */
  onClick: Function;
  /**
   * CSS id assigned to `<button>` tag.
   */
  id: string;
};

/**
 * Button used globally in site design.
 * @visibleName CTAButton
 */
const CTAButton: FunctionComponent<Props> = React.forwardRef(({
  id, color, text, onClick = () => '',
}: Props, ref: any) => (
  <Styled.Container ref={ref}>
    <Styled.Button id={id} color={color} onClick={onClick}>
      <Styled.ButtonText id={id}>
        {text}
      </Styled.ButtonText>
    </Styled.Button>
  </Styled.Container>
));

export default CTAButton;
