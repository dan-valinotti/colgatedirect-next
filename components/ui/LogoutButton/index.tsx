import React, { FunctionComponent, useEffect, useState } from 'react';
import { Button, CircularProgress } from '@material-ui/core';
import { withLogout } from '../../../hocs/withLogout';

/* LogoutButton component props */
interface Props {
  mutate: any;
}

/*
 * LogoutButton - component that performs a logout mutation on Shopify API when clicked
 * @param mutate: function (performs logout mutation)
 * @state accessToken: string (stores customerAccessToken from localStorage)
 * @state processing: boolean (true if running mutation)
 */
const LogoutButton: FunctionComponent<Props> = (
  { mutate }: Props,
) => {
  const [accessToken, setAccessToken] = useState<string>('');
  const [processing, setProcessing] = useState<boolean>(false);

  // Updates accessToken state var when client browser is detected
  useEffect(() => {
    if (window && window.localStorage) {
      setAccessToken(window.localStorage.getItem('customerAccessToken'));
    }
  }, []);

  // onClick function for Button component - performs logout mutation
  //  Success: reload page
  //  Failure: log error to console
  const submitLogoutRequest = () => {
    setProcessing(true);
    if (accessToken !== '') {
      mutate({
        variables: {
          customerAccessToken: accessToken,
        },
      })
        .then(() => {
          window.localStorage.removeItem('customerAccessToken');
          window.location.reload();
        })
        .catch((error) => {
          console.log(error);
          setProcessing(false);
        });
    }
  };

  return (
    <Button
      onClick={submitLogoutRequest}
      variant="outlined"
      color="secondary"
    >
      {!processing ? (
        <span>Log out</span>
      ) : (
        <CircularProgress
          variant="indeterminate"
          color="secondary"
          size={24}
        />
      )}
    </Button>
  );
};

export default withLogout(LogoutButton);
