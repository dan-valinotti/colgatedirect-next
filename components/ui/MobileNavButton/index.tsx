import React, { FunctionComponent, useState } from 'react';
import Link from 'next/link';
import { Styled } from './_styles';
import { TestInterface } from './_types';
import NavItems from '../NewNavBar/navItems.json';

type Props = {
  test?: string;
};

type ItemProps = {
  title: string;
  handle: string;
  products: any[];
};

const NavItem: FunctionComponent<ItemProps> = ({ title, handle, products }: ItemProps) => (
  <li>
    <Link href="/">
      {title}
    </Link>
    {products.length > 0 && (
      <ul>
        {products.map((subItem, key) => (
          <li key={key}>
            <Link href="/">
              {subItem.title}
            </Link>
          </li>
        ))}
      </ul>
    )}
  </li>
);

const MobileNavButton: FunctionComponent<Props> = ({ test }: Props) => {
  const [open, setOpen] = useState<boolean>(false);

  console.log(open);
  const toggleOpen = () => {
    setOpen(!open);
  };

  return (
    <Styled.Container>
      <Styled.IconContainer onClick={toggleOpen}>
        <i className="fas fa-bars" />
      </Styled.IconContainer>
      <Styled.NavWindowContainer open={open}>
        <Styled.NavWindow>
          <Styled.IconContainer onClick={toggleOpen}>
            <i className="fas fa-times" />
          </Styled.IconContainer>
          <Styled.NavItems>
            {NavItems.navigationItems.map((item, key) => (
              <NavItem
                title={item.title}
                handle={item.handle}
                products={item.products}
              />
            ))}
          </Styled.NavItems>
        </Styled.NavWindow>
      </Styled.NavWindowContainer>
    </Styled.Container>
  );
};

export default MobileNavButton;
