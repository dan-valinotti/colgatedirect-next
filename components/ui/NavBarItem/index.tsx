import React, { FunctionComponent, useState, useEffect } from 'react';
import Link from 'next/link';
import { Styled } from './_styles';

type SubItem = {
  title: string;
  handle: string;
};

type Props = {
  title: string;
  handle: string;
  products: SubItem[];
};

const NavBarItem: FunctionComponent<Props> = ({ title, handle, products }: Props) => {
  const [hover, setHover] = useState<boolean>(false);

  const handleMouseOver = () => {
    setHover(true);
  };
  const handleMouseLeave = () => {
    setHover(false);
  };

  return (
    <Styled.Container
      onFocus={handleMouseOver}
      onMouseOver={handleMouseOver}
      onMouseLeave={handleMouseLeave}
      className={handle !== '#' ? handle : title.toLowerCase().replace(' ', '-')}
    >
      <Link
        href={handle !== '#' ? { pathname: '/product', query: { handle } } : '#'}
        as={handle !== '#' ? `/products/${handle}` : '#'}
        passHref
      >
        <Styled.RootNavButton
          hovered={hover}
        >
          <h4>{title}</h4>
        </Styled.RootNavButton>
      </Link>
      {products.length > 0 && (
        <Styled.SubItemContainer
          onFocus={handleMouseOver}
          onMouseOver={handleMouseOver}
          onMouseLeave={handleMouseLeave}
          hovered={hover}
        >
          {products.map(({ title: subItemTitle, handle: subItemHandle }, key) => (
            <Link
              href={{ pathname: '/product', query: { handle: subItemHandle } }}
              as={`/products/${subItemHandle}`}
              passHref
              key={key}
            >
              <Styled.SubItem>
                <span>{subItemTitle}</span>
              </Styled.SubItem>
            </Link>
          ))}
        </Styled.SubItemContainer>
      )}
    </Styled.Container>
  );
};

export default NavBarItem;
