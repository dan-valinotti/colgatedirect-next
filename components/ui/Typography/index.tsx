import styled from 'styled-components';
import DynamicComponent from '../../DynamicComponent';
import theme from '../../../styles/theme';

const {
  heading,
  subheading,
  paragraph,
  heroHeading,
  heroSubheading,
  heroBody,
} = theme.textStyles;

const createTypoComponent = (props) => {
  const TypoComponent = styled(DynamicComponent)``;

  TypoComponent.defaultProps = {
    ...props,
  };

  return TypoComponent;
};

export const Heading = createTypoComponent(heading);
export const Subheading = createTypoComponent(subheading);
export const Paragraph = createTypoComponent(paragraph);
export const HeroHeading = createTypoComponent(heroHeading);
export const HeroSubheading = createTypoComponent(heroSubheading);
export const HeroBody = createTypoComponent(heroBody);
