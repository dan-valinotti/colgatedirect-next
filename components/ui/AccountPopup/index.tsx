import React, { FunctionComponent, useState, useEffect } from 'react';
import { useQuery } from '@apollo/react-hooks';
import Link from 'next/link';
import { Styled } from './_styles';
import LogoutButton from '../LogoutButton';
import { GetCustomerInfoResponse, CUSTOMER_INFO_QUERY } from '../../../common/queries/account';
import CTAButton from '../CTAButton';

const AccountPopup: FunctionComponent = () => {
  const [open, setOpen] = useState<boolean>(false);
  const [accessToken, setAccessToken] = useState<string>('');

  const {
    data: getCustomerData,
    error: getCustomerError,
    refetch: refetchCustomerData,
  } = useQuery<GetCustomerInfoResponse>(CUSTOMER_INFO_QUERY, {
    variables: {
      customerAccessToken: accessToken,
    },
    // Skip query execution if accessToken is not retrieved from
    //  localStorage, or if user is not logged in
    skip: (!accessToken || accessToken === ''),
  });

  // Handler functions for IconButton
  const toggleOpen = () => setOpen(!open);

  /**
   * useEffect() - monitor 'window' variable to wait for client browser to be
   *   detected, then retrieve 'customerAccessToken' from localStorage. If value
   *   is not empty, attempt to retrieve customer data
   */
  useEffect(() => {
    if (window && window.localStorage) {
      setAccessToken(window.localStorage.getItem('customerAccessToken'));
    }
    if (accessToken && accessToken !== '') {
      refetchCustomerData()
        .catch((err) => console.log(err));
    }
  }, [accessToken, refetchCustomerData]);

  return (
    <Styled.AccountButtonContainer>
      <i className="far fa-user-circle account-btn" role="presentation" onClick={toggleOpen} />
      <Styled.AccountPopupContainer open={open}>
        <Styled.AccountPopup id="popup-content">
          {getCustomerData ? (
            <Styled.AccountDetails>
              <h3>Account</h3>
              <p id="account-name">
                {getCustomerData.customer.firstName} {getCustomerData.customer.lastName}
                <br />
                {getCustomerData.customer.email}
              </p>
              <div>
                <CTAButton
                  id="account-page-btn"
                  color="primary"
                  text="Account details"
                  onClick={toggleOpen}
                />
                <LogoutButton />
              </div>
            </Styled.AccountDetails>
          ) : (
            <>
              <Link href="/login" passHref>
                <CTAButton
                  color="primary"
                  text="Log in"
                  onClick={toggleOpen}
                  id="login-btn"
                />
              </Link>
              <Link href="/register" passHref>
                <CTAButton
                  color="secondary"
                  text="Sign up"
                  onClick={toggleOpen}
                  id="signup-btn"
                />
              </Link>
            </>
          )}
        </Styled.AccountPopup>
      </Styled.AccountPopupContainer>
    </Styled.AccountButtonContainer>
  );
};

export default AccountPopup;
