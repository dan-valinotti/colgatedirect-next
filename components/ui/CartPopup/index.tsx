import React, { FunctionComponent, useState, useEffect } from 'react';
import { useQuery } from '@apollo/react-hooks';
import Link from 'next/link';
import { ApolloError } from 'apollo-boost';
import CartContent from 'components/CartContent';
import { Styled } from './_styles';
import LogoutButton from '../LogoutButton';
import CTAButton from '../CTAButton';
import { GetCartResponse } from './_types';

/**
 * properties
 */
interface Props {
  /** Cart data retrieved from Shopfy GraphQL API. */
  cart: GetCartResponse;
  clearCart: Function;
  total: number;
  getTotal: Function;
  createCartLoading: boolean;
  getCartLoading: boolean;
  createCartError: ApolloError;
  getCartError: ApolloError;
  getCartRefetch: Function;
  stopPolling: Function;
}

const CartPopup: FunctionComponent<Props> = ({
  cart, total, getTotal, clearCart, createCartLoading,
  getCartLoading, createCartError,
  getCartError, getCartRefetch,
  stopPolling,
}: Props) => {
  const [open, setOpen] = useState<boolean>(false);
  // Handler functions for IconButton
  const toggleOpen = () => {
    if (!open) {
      setOpen(true);
      getCartRefetch()
        .then(() => {
          if (!getCartLoading && !getCartError && cart) {
            // Recalculate total
            getTotal(cart.node.lineItems.edges);
          }
        })
        .catch((error) => console.log(error));
    } else {
      setOpen(false);
    }
  };

  useEffect(() => {
    if (cart) {
      getTotal(cart.node.lineItems.edges);
    }
  }, [cart, getTotal]);

  return (
    <Styled.CartButtonContainer>
      <i className="fas fa-shopping-cart" role="presentation" onClick={toggleOpen} />
      <Styled.CartPopupContainer open={open}>
        <Styled.CartPopup>
          {cart && (
            <CartContent
              cart={cart}
              total={total}
              clearCart={clearCart}
            />
          )}
        </Styled.CartPopup>
      </Styled.CartPopupContainer>
    </Styled.CartButtonContainer>
  );
};

export default CartPopup;
