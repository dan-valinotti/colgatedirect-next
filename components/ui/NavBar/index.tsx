import React, { FunctionComponent, useState } from 'react';
import Link from 'next/link';
import { Styled } from './_styles';
import NavItems from './navItems.json';
import NavBarItem from '../NavBarItem';
import NavIconButtons from '../NavIconButtons/index';
import MobileNavButton from '../MobileNavButton/index';

const NavBar: FunctionComponent = () => {
  const [mobileOpen, setMobileOpen] = useState<boolean>(false);

  const toggleMobileOpen = () => {
    setMobileOpen(!mobileOpen);
  };

  return (
    <Styled.Container id="navigation-bar">
      <Styled.MainContainer>
        <MobileNavButton onClick={toggleMobileOpen} />
        <Styled.LogoWrapper>
          <Styled.Logo>
            <Link href="/">
              <img id="nav-logo" src="/static/images/nav-logo.png" alt="ShopSmiles by Colgate" />
            </Link>
          </Styled.Logo>
        </Styled.LogoWrapper>
        <Styled.NavItemsContainer id="navigation-items">
          {NavItems.navigationItems.map(({ title, handle, products }, key) => (
            <NavBarItem
              title={title}
              handle={handle}
              products={products}
              key={key}
            />
          ))}
        </Styled.NavItemsContainer>
        <Styled.FlexContainer>
          <NavIconButtons mobileOpen={mobileOpen} />
        </Styled.FlexContainer>
      </Styled.MainContainer>
    </Styled.Container>
  );
};

export default NavBar;
