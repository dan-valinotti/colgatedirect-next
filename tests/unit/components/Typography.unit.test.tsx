/**
 * @jest-environment jsdom
*/
import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Heading, Subheading, Paragraph } from '../../../components/ui/Typography';

describe('<Heading />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <Heading as="h1" color="rgba(0,0,0,1)" textAlign="left">
        Test
      </Heading>
    );
  });

  it('Does the rendered component match the Snapshot?', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  })
  it('Does rendered component show correct text?', () => {
    expect(wrapper.text()).toMatch('Test');
  });
  it('Does component render correct HTML element?', () => {
    expect(wrapper.getDOMNode().tagName).toMatch('H1');
  })
});

describe('<Subheading />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <Subheading as="h6" color="rgba(0,0,0,1)" textAlign="left">
        Test
      </Subheading>
    );
  });

  it('Does the rendered component match the Snapshot?', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  })
  it('Does rendered component show correct text?', () => {
    expect(wrapper.text()).toMatch('Test');
  });
  it('Does component render correct HTML element?', () => {
    expect(wrapper.getDOMNode().tagName).toMatch('H6');
  })
});

describe('<Paragraph />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <Paragraph as="p" color="rgba(0,0,0,1)" textAlign="left">
        Test
      </Paragraph>
    );
  });

  it('Does the rendered component match the Snapshot?', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  })
  it('Does rendered component show correct text?', () => {
    expect(wrapper.text()).toMatch('Test');
  });
  it('Does component render correct HTML element?', () => {
    expect(wrapper.getDOMNode().tagName).toMatch('P');
  })
});
