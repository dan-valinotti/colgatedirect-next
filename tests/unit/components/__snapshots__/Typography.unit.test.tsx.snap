// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> Does the rendered component match the Snapshot? 1`] = `
<Styled(styled.p)
  as="h1"
  color="rgba(0,0,0,1)"
  fontFamily="\\"Colgate Ready\\", serif"
  fontSize={
    Array [
      "1.75em",
      "1.75em",
      "1.75em",
      "1.75em",
      "2em",
    ]
  }
  fontWeight={700}
  textAlign="left"
>
  <StyledComponent
    as="h1"
    color="rgba(0,0,0,1)"
    fontFamily="\\"Colgate Ready\\", serif"
    fontSize={
      Array [
        "1.75em",
        "1.75em",
        "1.75em",
        "1.75em",
        "2em",
      ]
    }
    fontWeight={700}
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "_foldedDefaultProps": Object {
          "color": "rgba(0,0,0,0.8)",
          "fontFamily": "\\"Colgate Ready\\", serif",
          "fontSize": Array [
            "1.75em",
            "1.75em",
            "1.75em",
            "1.75em",
            "2em",
          ],
          "fontWeight": 700,
        },
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bwzfXH",
          "isStatic": false,
          "lastClassName": "OkwrY",
          "rules": Array [
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
",
          ],
        },
        "displayName": "Styled(styled.p)",
        "foldedComponentIds": Array [
          "sc-bdVaJa",
        ],
        "render": [Function],
        "styledComponentId": "sc-bwzfXH",
        "target": "p",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    textAlign="left"
  >
    <h1
      className="sc-bdVaJa sc-bwzfXH OkwrY"
      color="rgba(0,0,0,1)"
      fontFamily="\\"Colgate Ready\\", serif"
      fontSize={
        Array [
          "1.75em",
          "1.75em",
          "1.75em",
          "1.75em",
          "2em",
        ]
      }
      fontWeight={700}
    >
      Test
    </h1>
  </StyledComponent>
</Styled(styled.p)>
`;

exports[`<Paragraph /> Does the rendered component match the Snapshot? 1`] = `
<Styled(styled.p)
  as="p"
  color="rgba(0,0,0,1)"
  fontFamily="\\"Colgate Ready\\", serif"
  fontSize={
    Array [
      "0.9em",
      "1em",
    ]
  }
  fontWeight={400}
  textAlign="left"
>
  <StyledComponent
    as="p"
    color="rgba(0,0,0,1)"
    fontFamily="\\"Colgate Ready\\", serif"
    fontSize={
      Array [
        "0.9em",
        "1em",
      ]
    }
    fontWeight={400}
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "_foldedDefaultProps": Object {
          "color": "rgba(0,0,0,0.8)",
          "fontFamily": "\\"Colgate Ready\\", serif",
          "fontSize": Array [
            "0.9em",
            "1em",
          ],
          "fontWeight": 400,
        },
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bxivhb",
          "isStatic": false,
          "lastClassName": "beLkIv",
          "rules": Array [
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
",
          ],
        },
        "displayName": "Styled(styled.p)",
        "foldedComponentIds": Array [
          "sc-bdVaJa",
        ],
        "render": [Function],
        "styledComponentId": "sc-bxivhb",
        "target": "p",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    textAlign="left"
  >
    <p
      className="sc-bdVaJa sc-bxivhb beLkIv"
      color="rgba(0,0,0,1)"
      fontFamily="\\"Colgate Ready\\", serif"
      fontSize={
        Array [
          "0.9em",
          "1em",
        ]
      }
      fontWeight={400}
    >
      Test
    </p>
  </StyledComponent>
</Styled(styled.p)>
`;

exports[`<Subheading /> Does the rendered component match the Snapshot? 1`] = `
<Styled(styled.p)
  as="h6"
  color="rgba(0,0,0,1)"
  fontFamily="\\"Colgate Ready\\", serif"
  fontSize={
    Array [
      "1.25em",
      "1.25em",
      "1.25em",
      "1.25em",
      "1.75em",
    ]
  }
  fontWeight={400}
  textAlign="left"
>
  <StyledComponent
    as="h6"
    color="rgba(0,0,0,1)"
    fontFamily="\\"Colgate Ready\\", serif"
    fontSize={
      Array [
        "1.25em",
        "1.25em",
        "1.25em",
        "1.25em",
        "1.75em",
      ]
    }
    fontWeight={400}
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "_foldedDefaultProps": Object {
          "color": "rgba(0,0,0,0.6)",
          "fontFamily": "\\"Colgate Ready\\", serif",
          "fontSize": Array [
            "1.25em",
            "1.25em",
            "1.25em",
            "1.25em",
            "1.75em",
          ],
          "fontWeight": 400,
        },
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-htpNat",
          "isStatic": false,
          "lastClassName": "euwnsm",
          "rules": Array [
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
  ",
            [Function],
            "
",
          ],
        },
        "displayName": "Styled(styled.p)",
        "foldedComponentIds": Array [
          "sc-bdVaJa",
        ],
        "render": [Function],
        "styledComponentId": "sc-htpNat",
        "target": "p",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    textAlign="left"
  >
    <h6
      className="sc-bdVaJa sc-htpNat euwnsm"
      color="rgba(0,0,0,1)"
      fontFamily="\\"Colgate Ready\\", serif"
      fontSize={
        Array [
          "1.25em",
          "1.25em",
          "1.25em",
          "1.25em",
          "1.75em",
        ]
      }
      fontWeight={400}
    >
      Test
    </h6>
  </StyledComponent>
</Styled(styled.p)>
`;
